datalad/credman.py

```diff
@@ -0,0 +1,621 @@
+    system. All properties are stored in the `global` (i.e., user) scope
+    under configuration items following the pattern
+
+      `datalad.credential.<name>.<property>`
+
+    where `<name>` is a credential name/identifier, and `<property>` is an
+    arbitrarily named credential property, whose name must follow the
+    git-config syntax for variable names (case-insensitive, only alphanumeric
+    characters and -, and must start with an alphabetic character).
+
+    Create a `CredentialManager` instance is fast, virtually no initialization
+    needs to be performed. All internal properties are lazily evaluated.
+    This facilitate usage of this facility in code where it is difficult
+    to incorporate a long-lived central instance.
+
+    API
+
+    With one exception, all parameter names of methods in the main API
+    outside **kwargs must have a `_` prefix that distinguishes credential
+    properties from method parameters. The one exception is the `name`
+    parameter, which is used as a primary identifier (albeit being optional
+    for some operations).
+    """
+    def __init__(self, cfg=None):
+        """
+
+        Parameters
+        ----------
+        cfg: ConfigManager, optional
+          If given, all configuration queries are performed using this
+          ConfigManager instance. Otherwise `datalad.cfg` is used.
+        """
+        self.__cfg = cfg
+        self.__cred_types = None
+        self.__keyring = None
+
+    # main API
+    #
+    def get(self, name=None, _prompt=None, _type_hint=None, **kwargs):
+        """Get properties and secret of a credential.
+
+        This is a read-only method that never modifies information stored
+        on a credential in any backend.
+
+        Credential property lookup is supported via a number approaches.  When
+        providing `name`, all existing corresponding configuration items are
+        found and reported, and an existing secret is retrieved from name-based
+        secret backends (presently `keyring`). When providing a `type` property
+        or a `_type_hint` the lookup of additional properties in the
+        keyring-backend is enabled, using predefined property name lists
+        for a number of known credential types.
+
+        For all given property keys that have no value assigned after the
+        initial lookup, manual/interactive entry is attempted, whenever
+        a custom `_prompt` was provided. This include requesting a secret.
+        If manually entered information is contained in the return credential
+        record, the record contains an additional `_edited` property with a
+        value of `True`.
+
+        If no secret is known after lookup and a potential manual data entry,
+        a plain `None` is returned instead of a full credential record.
+
+        Parameters
+        ----------
+        name: str, optional
+          Name of the credential to be retrieved
+        _prompt: str or None
```

Comment:
Not exactly essential, but seems like `_prompt` should be a dict `{<name of missing property to ask for>: <string to use for prompting>}`, no?

Reply:
Can you elaborate on why that is needed?

Reply:
> For all given property keys that have no value assigned after the
> initial lookup, manual/interactive entry is attempted, whenever
> a custom `_prompt` was provided.

If there's only one string, then all those property keys can only be prompted for by the exact same question it seems. If this is not supposed to only cover a missing `secret`, then several may be needed.

Reply:
Do you have any example where this would be needed? The current system also has a single one and seems fine.